### ISSUE TITLE
libcurl sometimes hangs with socks5h:// proxies

### ISSUE THREAD

**Issue by Snawoot** [curl/curl#5710]
### I did this

```
curl -v --proxy socks5h://127.0.0.1:1111 2ip.ru
```

### I expected the following

```
*   Trying 127.0.0.1:1111...
* SOCKS5 connect to 2ip.ru:80 (remotely resolved)
* SOCKS5 request granted.
* Connected to 127.0.0.1 (127.0.0.1) port 1111 (#0)
> GET / HTTP/1.1
> Host: 2ip.ru
> User-Agent: curl/7.69.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 22 Jul 2020 16:05:32 GMT
< Content-Type: application/octet-stream
< Content-Length: 15
< Connection: keep-alive
< 
159.203.173.33
```

instead, sometimes curl hangs:

```
*   Trying 127.0.0.1:1111...
* SOCKS5 connect to 2ip.ru:80 (remotely resolved)
```
### Additional info

It seems, libcurl enters into some race condition if SOCKS5 proxy responds too fast. I can confirm with strace SOCKS5 connection was granted, but libcurl stuck in epoll loop:

```
4950  write(2, "  Trying 127.0.0.1:1111...\n", 27) = 27
4950  setsockopt(5, SOL_TCP, TCP_NODELAY, [1], 4) = 0
4950  setsockopt(5, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
4950  setsockopt(5, SOL_TCP, TCP_KEEPIDLE, [60], 4) = 0
4950  setsockopt(5, SOL_TCP, TCP_KEEPINTVL, [60], 4) = 0
4950  fcntl(5, F_GETFL)                 = 0x2 (flags O_RDWR)
4950  fcntl(5, F_SETFL, O_RDWR|O_NONBLOCK) = 0
4950  connect(5, {sa_family=AF_INET, sin_port=htons(1111), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Операция выполняется в данный момент)
4950  poll([{fd=5, events=POLLOUT|POLLWRNORM}], 1, 0) = 1 ([{fd=5, revents=POLLOUT|POLLWRNORM}])
4950  getsockopt(5, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
4950  sendto(5, "\5\2\0\1", 4, MSG_NOSIGNAL, NULL, 0) = 4
4950  recvfrom(5, "\5\0", 2, 0, NULL, NULL) = 2
4950  write(2, "*", 1)                  = 1
4950  write(2, " ", 1)                  = 1
4950  write(2, "SOCKS5 connect to 2ip.ru:80 (remotely resolved)\n", 48) = 48
4950  sendto(5, "\5\1\0\3\0062ip.ru\0P", 13, MSG_NOSIGNAL, NULL, 0) = 13
4950  recvfrom(5, "\5\0\0\1\0\0\0\0\0\0", 10, 0, NULL, NULL) = 10
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 199) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = 0 (Timeout)
4950  rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, 8) = 0
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  recvfrom(5, 0x560efc089806, 3, 0, NULL, NULL) = -1 EAGAIN (Ресурс временно недоступен)
4950  rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[PIPE], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f3123e58a90}, NULL, 8) = 0
4950  poll([{fd=5, events=POLLIN}, {fd=3, events=POLLIN}], 2, 1000) = ? ERESTART_RESTARTBLOCK (Interrupted by signal)
4950  --- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
4950  +++ killed by SIGINT +++
```

As you can see from `4950  recvfrom(5, "\5\0\0\1\0\0\0\0\0\0", 10, 0, NULL, NULL) = 10` connection was granted by SOCKS5 server.

GDB stack trace:

```
(gdb) bt
#0  0x00007ffff7af5a77 in poll () from /lib64/libc.so.6
#1  0x00007ffff7f45bcc in poll (__timeout=1220417006, __nfds=0, __fds=0x7fffffffda90) at /usr/include/bits/poll2.h:46
#2  Curl_poll (ufds=ufds@entry=0x7fffffffda90, nfds=nfds@entry=2, timeout_ms=timeout_ms@entry=1000) at ../../lib/select.c:380
#3  0x00007ffff7f3f52a in Curl_multi_wait (multi=multi@entry=0x5555555b20b0, extra_fds=extra_fds@entry=0x0, extra_nfds=extra_nfds@entry=0, timeout_ms=timeout_ms@entry=1000, ret=ret@entry=0x0, extrawait=extrawait@entry=true, 
    use_wakeup=true) at ../../lib/multi.c:1172
#4  0x00007ffff7f3f6ba in Curl_multi_wait (use_wakeup=true, extrawait=true, ret=0x0, timeout_ms=1000, extra_nfds=0, extra_fds=0x0, multi=0x5555555b20b0) at ../../lib/multi.c:1060
#5  curl_multi_poll (multi=multi@entry=0x5555555b20b0, extra_fds=extra_fds@entry=0x0, extra_nfds=extra_nfds@entry=0, timeout_ms=timeout_ms@entry=1000, ret=ret@entry=0x0) at ../../lib/multi.c:1261
#6  0x00007ffff7f37cb0 in easy_transfer (multi=<optimized out>) at ../../lib/easy.c:599
#7  easy_perform (events=false, data=0x5555555b9af0) at ../../lib/easy.c:691
#8  curl_easy_perform (data=0x5555555b9af0) at ../../lib/easy.c:710
#9  0x000055555556ed37 in serial_transfers (share=0x5555555b1ea0, global=0x7fffffffdd20) at ../../src/tool_operate.c:2203
#10 run_all_transfers (result=CURLE_OK, share=0x5555555b1ea0, global=0x7fffffffdd20) at ../../src/tool_operate.c:2379
#11 operate (global=0x7fffffffdd20, argc=<optimized out>, argv=<optimized out>) at ../../src/tool_operate.c:2491
#12 0x000055555555f1fe in main (argc=5, argv=0x7fffffffdeb8) at ../../src/tool_main.c:317
(gdb) 
```

It's worth noting I can't reproduce this issue with `socks5://` proxy URL (w/o remote resolve). It occurs only with `socks5h://` proxy scheme. My SOCKS5 proxy server is a local shadowsocks instance (tried both go-shadowsocks2 and shadowsocks-libev with the same result).

### curl/libcurl version

```
curl 7.69.1 (x86_64-redhat-linux-gnu) libcurl/7.69.1 OpenSSL/1.1.1g-fips zlib/1.2.11 brotli/1.0.7 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh/0.9.4/openssl/zlib nghttp2/1.41.0
Release-Date: 2020-03-11
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz Metalink NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

### operating system

Fedora 32:

```
Linux holy-see 5.7.9-200.fc32.x86_64 #1 SMP Fri Jul 17 16:23:37 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by bagder** [comment]
Please retry with 7.71.0 as we fixed some socks connection regressions in that release.

**Comment by Snawoot** [comment]
I've can confirm it's fixed in 7.71.0. I've tested docker images from 7.69.1 to 7.71.1. Issue stops from reproducing on 7.71.0.
